lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
       @namespace = namespace
       @context = context
-      @file = file
       @logger = logger
+      @template = template
     end
 
     def id
       "#{type}/#{name}"
     end
 
+    def file
```

Comment:
This method would return an instance of closed `Tempfile`:

```
#<Tempfile:/var/folders/9k/zxhywlyx2nx_3z3qnrzpkjd00000gn/T/abc20170623-92852-1180pr4 (closed)>
```

That we'd be unable to read (because read of closed file raises with `IOError: closed stream`)

I don't have full context on how you're using the return value of this method, just letting you know that it may not work as expected.